File: src/deno-addons/queue.js
const libQueue = `./libqueue.so`;

const queuelib = Deno.dlopen(
    libQueue,
    {
        "queue_new": { parameters: [], result: "pointer" },
        "queue_push": { parameters: ["pointer", "i32"], result: "i32" },
        "queue_pop": { parameters: ["pointer"], result: "i32" },
        "queue_peek": { parameters: ["pointer"], result: "i32" },
        "queue_is_empty": { parameters: ["pointer"], result: "bool" },
    }
);

export class Queue {
    constructor() {
        this.ptr = queuelib.symbols.queue_new();
    }

    push(item) {
        return queuelib.symbols.queue_push(this.ptr, item);
    }

    pop() {
        return queuelib.symbols.queue_pop(this.ptr);
    }

    peek() {
        return queuelib.symbols.queue_peek(this.ptr);
    }

    isEmpty() {
        return queuelib.symbols.queue_is_empty(this.ptr);
    }

}

File: src/wasm/queue.cpp
#include <emscripten/bind.h>
#include <queue>

using namespace emscripten;

class Queue {
private:
  std::queue<int> items;

public:
  Queue(){ }
  int push(int item);
  int pop();
  int peek();
  bool isEmpty();
  // std::queue<int> getItems();
};


int Queue::push(int item) {
  this->items.push(item);
  return this->items.size() - 1;
}

int Queue::pop() {
  int item = this->items.front();
  this->items.pop();
  return item;
}

int Queue::peek() { return this->items.front(); }
bool Queue::isEmpty() { return this->items.empty(); }

EMSCRIPTEN_BINDINGS(cola) {
  class_<Queue>("Queue")
    .constructor<>()
    .function("push", &Queue::push)
    .function("pop", &Queue::pop)
    .function("peek", &Queue::peek)
    .function("isEmpty", &Queue::isEmpty);
}